src/server/saved_objects/saved_objects_mixin.js

```diff
@@ -26,11 +26,30 @@ export function savedObjectsMixin(kbnServer, server) {
   server.route(createGetRoute(prereqs));
   server.route(createUpdateRoute(prereqs));
 
+  async function onBeforeWrite() {
+    const adminCluster = server.plugins.elasticsearch.getCluster('admin');
+
+    try {
+      await adminCluster.callWithInternalUser('cluster.health', {
+        timeout: `${server.config().get('savedObjects.indexCheckTimeout')}ms`,
+        index: server.config().get('kibana.index'),
+        waitForStatus: 'yellow',
+      });
+    } catch (error) {
+      if (error && error.body && error.body.status === 'red') {
+        server.log(['debug', 'savedObjects'], `Attempted to write to the Kibana index when it didn't exist.`);
```

Comment:
Is it possible the status is red for a reason other than the index not existing?

Reply:
Yeah, I suppose it is. I'll update the message.

Reply:
I did a little digging and confirmed that we can detect when a red status is because there are uninitialized or otherwise unavailable shards. In those cases we will no longer throw a 404, instead we will let the request fail on its own and provide its own descriptive error.

red status when all data nodes are gone but index does exist
![image](https://user-images.githubusercontent.com/1329312/31040559-99f92dd2-a53d-11e7-90cd-c6fd4b5c7734.png)

red status when es is healthy but Kibana index doesn't exist
![image](https://user-images.githubusercontent.com/1329312/31040543-55cf0960-a53d-11e7-870e-1ef4e9968b76.png)